File: Leetcode/572.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func max(a ...int) int {
	res := a[0]
	for i := 0; i < len(a); i++ {
		if a[i] > res {
			res = a[i]
		}
	}
	return res
}

func dfs(node *TreeNode, res *[][]int) int {
	if node == nil {
		return 0
	}
	left := dfs(node.Left, res)
	right := dfs(node.Right, res)
	height := max(left, right) + 1
	*res = append(*res, []int{node.Val, height})
	return height
}

func compareData(a1 []int, a2 []int) bool {
	return a1[0] == a2[0] && a1[1] == a2[1]
}

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	rootValues := make([][]int, 0)
	subRootValues := make([][]int, 0)

	dfs(root, &rootValues)
	dfs(subRoot, &subRootValues)

	last := len(subRootValues) - 1
	mid := (last + 1) / 2
	j := 0
	for i := 0; i <= len(rootValues)-last; i++ {
		if compareData(rootValues[i], subRootValues[j]) && compareData(rootValues[i+mid], subRootValues[mid]) && compareData(rootValues[i+last], subRootValues[last]) {
			ok := true
			for k := 0; k <= last; k++ {
				if !compareData(rootValues[i+k], subRootValues[k]) {
					ok = false
					break
				}
			}
			if ok {
				return true
			}
		}
	}
	return false
}

File: Leetcode/572.py
# Definition for a binary tree node.
import hashlib
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def hash(value) ->str:
            return hashlib.md5(value.encode()).hexdigest()
        
        def get_hash(node: TreeNode) -> str:
            if node is None:
                return hash("None")
            return hash(str(node.val)  + get_hash(node.left) + get_hash(node.right))
        
        dst_value = get_hash(subRoot)
        def get_hash_and_compare(node):
            if node is None:
                return hash("None"), False
            left_hash,  found = get_hash_and_compare(node.left)
            if found:
                return "", True
            right_hash,  found = get_hash_and_compare(node.right)
            if found:
                return "", True
            hash_value = hash(str(node.val)  + left_hash + right_hash)
            if hash_value == dst_value:
                return "", True
            return hash_value, False
        
        return get_hash_and_compare(root)[1]
            